query/src/main/scala/filodb/query/exec/SetOperatorExec.scala

```diff
@@ -375,31 +376,107 @@ final case class SetOperatorExec(queryContext: QueryContext,
   /***
    * Return LHS range vector which have join keys not present in RHS
    */
-  private[exec] def setOpUnless(lhsRvs: List[RangeVector],
-                          rhsRvs: List[RangeVector]): Iterator[RangeVector] = {
+  // scalastyle:off method.length
+  private[exec] def setOpUnless(
+                                 lhsRvs: List[RangeVector],
+                                 rhsRvs: List[RangeVector],
+                                 rhsSchema: ResultSchema,
+                                 querySession: QuerySession
+                               ): Iterator[RangeVector] = {
+
+    // 1) Prep the serializer
+    lazy val builder      = SerializedRangeVector.newBuilder()
+    lazy val recSchema    = SerializedRangeVector.toSchema(rhsSchema.columns, rhsSchema.brSchemas)
+    lazy val execPlanName = queryWithPlanName(queryContext)
 
-    val rhsKeysSet = new mutable.HashSet[Map[Utf8Str, Utf8Str]]()
-    rhsRvs.foreach { rv =>
+    // 2) Turn every RHS into a SerializedRangeVector
+    val mappedRhs = rhsRvs.map {
+      case srv: SerializedRangeVector => srv
+      case rv: RangeVector =>
+        SerializedRangeVector(rv, builder, recSchema, execPlanName, querySession.queryStats)
+    }
+
+    // 3) Stitch shards of RHS into a single RV per join-key
+    val rhsMap  = new mutable.HashMap[Map[Utf8Str, Utf8Str], RangeVector]()
+    mappedRhs.foreach { rv =>
       val jk = joinKeys(rv.key)
-      rhsKeysSet += jk
+      if (!isEmpty(rv, rhsSchema)) {
+        if (rhsMap.contains(jk)) {
+          val resVal = rhsMap(jk)
+          if (resVal.key.labelValues == rv.key.labelValues) {
+            rhsMap.put(jk, StitchRvsExec.stitch(rv, resVal, outputRvRange))
+          } else {
+            rhsMap.put(jk, rv)
+          }
+        } else rhsMap.put(jk, rv)
+      }
     }
-    // Add range vectors which are not present in rhs
-
-    val result = lhsRvs.foldLeft(
-      mutable.Map.empty[Map[Utf8Str, Utf8Str], mutable.Map[Map[Utf8Str, Utf8Str], RangeVector]]) {
-      case (accumulated, lhs) =>
-        val jk = joinKeys(lhs.key)
-        if (!rhsKeysSet.contains(jk)) {
-          val result = accumulated.getOrElseUpdate(jk, mutable.Map.empty)
-          val stitchedOrNewValue = result.get(lhs.key.labelValues) match {
-            case Some(matched)  => StitchRvsExec.stitch(lhs, matched, outputRvRange)
-            case None           => lhs
+
+    // 4) Prepare result accumulator
+    val result = new mutable.HashMap[Map[Utf8Str, Utf8Str], ArrayBuffer[RangeVector]]()
+    val period = lhsRvs.headOption.flatMap(_.outputRange)
+
+    // 5) For each LHS, either emit unchanged (if no RHS) or do the “unless”‐zip
+    lhsRvs.foreach { lhs =>
+      val jk = joinKeys(lhs.key)
+
+      if (rhsMap.contains(jk)) {
+        var index = -1
+        val lhsStitched =
+          if (result.contains(jk)) {
```

Comment:
When will the result contain LHS join keys? Er, more precisely: when will the result ever contain something other than keys as-is in from the LHS?

Reply:
OK, I see. `result` is the map of join keys to matching `RangeVector`s.
The reason it's needed is for a fast lookup of a matching series (if one exists) so we can `stitch`? I.e. the reason `result` is needed is because we need to handle possibly repeat series labels (not join-key labels)?

If that's the purpose: should we consider something like a hash-map? Worst-case scenario: all LHS keys have the same join key. Would that mean there's a linear scan of all series every time we invoke `resVal.indexWhere` to add a `RangeVector` to `result`?